finalfusion-ffi/src/lib.rs

```diff
@@ -0,0 +1,67 @@
+use std::cell::RefCell;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::ffi::OsStr;
+use std::fs::File;
+use std::io::BufReader;
+use std::os::raw::c_char;
+use std::os::unix::ffi::OsStrExt;
+use std::ptr;
+
+use finalfusion::prelude::*;
+
+thread_local! {
+    static ERROR: RefCell<CString> = RefCell::new(CString::new(Vec::new()).unwrap());
+}
+
+fn update_error(err: &str) {
+    ERROR
+        .with(|e| e.replace(CString::new(err.as_bytes()).expect("Error string contains nul byte")));
+}
+
+/// Return the last error message.
+///
+/// The returned pointer is owned by the library and is valid until
+/// the next function call that can result in an error.
+#[no_mangle]
+pub extern "C" fn ff_error() -> *const c_char {
+    ERROR.with(|f| f.borrow().as_ptr())
+}
+
+#[no_mangle]
+pub extern "C" fn ff_read_embeddings(
+    filename: *const c_char,
+) -> *mut Embeddings<VocabWrap, StorageWrap> {
+    let filename = unsafe { CStr::from_ptr(filename) };
```

Comment:
Is `CStr::from_ptr` save to call on a null pointer?

Reply:
Nope, there should be an assertion before this. Will add...

Reply:
Note though that not all software does null pointer checks. The caller could also provide a wild pointer or other invalid pointer and there would be no way to check.

Reply:
Sure, I thought null checks are cheap enough and it might save someone some headaches if we are verbose about that error.